kernel/src/path.rs

```diff
@@ -1,34 +1,94 @@
 //! Utilities to make working with directory and file paths easier
 
-use lazy_static::lazy_static;
-use regex::Regex;
 use url::Url;
 
 use crate::{DeltaResult, Version};
 
-/// The delimiter to separate object namespaces, creating a directory structure.
+/// The delimiter to separate object namespaces, creating a directory structure. Note this is in url
+/// terms, so we use `/`
 const DELIMITER: &str = "/";
 
-lazy_static! {
-    static ref CHECKPOINT_FILE_PATTERN: Regex =
-        Regex::new(r"\d+\.checkpoint(\.\d+\.\d+)?\.parquet").unwrap();
-    static ref DELTA_FILE_PATTERN: Regex = Regex::new(r"\d+\.json").unwrap();
+#[derive(Debug)]
+pub(crate) struct LogPath<'a> {
+    url: &'a Url,
+    path: &'a str,
+    version: Option<Version>,
+    is_commit: bool,
+    is_checkpoint: bool,
 }
 
-#[derive(Debug)]
-pub(crate) struct LogPath<'a>(pub(crate) &'a Url);
+fn get_filename(path: &str) -> Option<&str> {
+    if path.is_empty() || path.ends_with('/') {
+        None
+    } else {
+        path.rsplit(DELIMITER).next()
+    }
+}
+
+fn try_get_version(version_str_opt: Option<&str>, expected_digits: usize) -> Option<Version> {
+    version_str_opt.and_then(|version_str| {
+        if version_str.len() == expected_digits {
+            version_str.parse().ok()
+        } else {
+            None
+        }
+    })
+}
+
+pub(crate) fn version_from_location(location: &Url) -> Option<Version> {
+    let path = location.path();
+    get_filename(path)
+        .and_then(|f| f.split_once('.'))
+        .and_then(|(name, _)| try_get_version(Some(name), 20))
+}
 
 impl<'a> LogPath<'a> {
+    pub(crate) fn new(url: &'a Url) -> Self {
+        let path = url.path();
+        let version_str = get_filename(path).and_then(|f| f.split_once('.'));
+        let version = version_str.and_then(|(name, _)| try_get_version(Some(name), 20));
+
+        let mut is_commit = false;
+        let mut is_checkpoint = false;
+        if version.is_some() {
+            // could be a checkpoint or commit file, let's check
+            let (_, suffix) = version_str.unwrap(); // safe, version.is_some()
+            is_commit = suffix == "json"; // if we were just [version].json, we're a commit file
```

Comment:
Should we take the opportunity to also parse [compaction files](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#log-compaction-files) - a thing I was unaware of when we wrote the initial `LogPath`.

Compaction files will look like: `00000000000000000004.00000000000000000006.compacted.json` which covers commits 4-6. They are optional to handle and just an optimization. Since I believe we would not do anything wrong, we could also just track that as an issue.

Reply:
Nice yeah, I've added that support